kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -0,0 +1,386 @@
+            return additionalHeaders;
+        }
+        Header[] newHeaders = new Header[oldHeaders.length + additionalHeaders.length];
+        System.arraycopy(additionalHeaders, 0, newHeaders, 0, additionalHeaders.length);
+        System.arraycopy(oldHeaders, 0, newHeaders, additionalHeaders.length, oldHeaders.length);
+        return newHeaders;
+    }
+
+    @NonNull
+    static Header[] removeInitialHeaders(@NonNull Header[] oldHeaders, int numToRemove) {
+        if (numToRemove < 0) {
+            throw new IllegalArgumentException();
+        }
+        Header[] newHeaders = new Header[oldHeaders.length - numToRemove];
+        if (newHeaders.length > 0) {
+            System.arraycopy(oldHeaders, numToRemove, newHeaders, 0, newHeaders.length);
+        }
+        return newHeaders;
+    }
+
+    static Set<RecordField> encryptedFields(Record kafkaRecord) {
+        for (Header header : kafkaRecord.headers()) {
+            if (FIELDS_HEADER_NAME.equals(header.key())) {
+                return RecordField.fromBits(header.value()[0]);
+            }
+        }
+        return EnumSet.noneOf(RecordField.class);
+    }
+
+    static RecordHeader dek(Record kafkaRecord) {
+        for (Header header : kafkaRecord.headers()) {
+            if (DEK_HEADER_NAME.equals(header.key())) {
+                return (RecordHeader) header;
+            }
+        }
+        throw new IllegalStateException();
+    }
+
+    private CompletionStage<AesGcmEncryptor> getOrCacheDecryptor(RecordHeader dekHeader,
+                                                                 K kekId,
+                                                                 E edek) {
+        return decryptorCache.compute(dekHeader, (k, v) -> {
+            if (v == null) {
+                return kms.decryptEdek(kekId, edek)
+                        .thenApply(AesGcmEncryptor::forDecrypt).toCompletableFuture();
+                // TODO what happens if the CS complete exceptionally
+                // TODO what happens if the CS doesn't complete at all in a reasonably time frame?
+            }
+            else {
+                return v;
+            }
+        });
+    }
+
+    @NonNull
+    @Override
+    public CompletionStage<Void> decrypt(@NonNull List<? extends Record> records,
+                                         @NonNull Receiver receiver) {
+        List<CompletionStage<Void>> futures = new ArrayList<>(records.size());
+        for (var kafkaRecord : records) {
+            var encryptedFields = encryptedFields(kafkaRecord);
+            if (encryptedFields.isEmpty()) {
+                receiver.accept(kafkaRecord, kafkaRecord.value(), kafkaRecord.headers());
+                futures.add(CompletableFuture.completedFuture(null));
+            }
+            else {
+                // right now (because we only support topic name based kek selection) noce we've resolve the first value we
+                // can keep the lock and process all the records
```

Comment:
is this comment accurate? It doesn't look like we take a lock to process all records but synchronize on the encryptor within decryptRecord